### PR TITLE
SONARJAVA-4829: Fix FP in S2694 on local classes

### PR DESCRIPTION
**Pull request by ADarko22**
The changes are split into 2 commits:

1. the first commit has the fix that avoids reporting in the case of local classes
2. the second commit is a rework to improve the readability and understanding of the implementation logic. It separates the state of the rule, i.e. the logic that detects if a nested class references an outer class into a nested static class.

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/InnerStaticClassesCheck.java [hunk 1]
```diff
@@ -52,131 +52,84 @@ public void scanFile(final JavaFileScannerContext context) {
...
+    state.updateOuterClassesReferencedBy(tree.symbol());
   }
 
   @Override
   public void visitNewClass(NewClassTree tree) {
     super.visitNewClass(tree);
-    checkSymbol(tree.symbolType().symbol());
-  }
-
-  private void checkSymbol(Symbol symbol) {
```
Comment (ADarko22) [inline java-checks/src/main/java/org/sonar/java/checks/InnerStaticClassesCheck.java:133]: moved to `State` as `updateOuterClassesReferencedBy(symbol)`

File: java-checks/src/main/java/org/sonar/java/checks/InnerStaticClassesCheck.java [hunk 2]
```diff
@@ -52,131 +52,84 @@ public void scanFile(final JavaFileScannerContext context) {
...
-          atLeastOneReference.pop();
-        }
-        while (atLeastOneReference.size() != outerClasses.size()) {
-          atLeastOneReference.push(Boolean.TRUE);
-        }
-      }
-    }
-  }
-
-  private int referenceInstance(Symbol symbol) {
```
Comment (ADarko22) [inline java-checks/src/main/java/org/sonar/java/checks/InnerStaticClassesCheck.java:147]: moved to `State` as `findReferencedClassIndex(symbol)`

File: java-checks/src/main/java/org/sonar/java/checks/InnerStaticClassesCheck.java [hunk 3]
```diff
@@ -52,131 +52,84 @@ public void scanFile(final JavaFileScannerContext context) {
...
-      if (symbol.isUnknown()) {
-        result = atLeastOneReference.size() - 1;
-      } else if (!symbol.isStatic()) {
-        result = fromInstance(symbol, owner);
-      }
-    }
-    return result;
-  }
-
-  private int fromInstance(Symbol symbol, Symbol owner) {
```
Comment (ADarko22) [inline java-checks/src/main/java/org/sonar/java/checks/InnerStaticClassesCheck.java:164]: moved to `State` as `findReferencedClassIndex(symbol, owner)`

**Review by johann-beleites-sonarsource — APPROVED** [review]
LGTM! Try rebasing on master for getting the ruling tests stable again.

**Pull request merged**